src/widget/woverview.h

```diff
@@ -99,9 +101,18 @@ class WOverview : public WWidget, public TrackDropTarget {
     void slotWaveformSummaryUpdated();
 
   private:
-    // Append the waveform overview pixmap according to available data in waveform
+    // Append the waveform overview pixmap according to available data
+    // in waveform
     virtual bool drawNextPixmapPart() = 0;
-    void paintText(const QString &text, QPainter *painter);
+    void drawEndOfTrackBackground(ScopedPainter painter);
```

Comment:
It's not obvious why the ScopedPainter is passed by value here. Someone might change it to a const-ref to improve efficiency. But then the behavior would be completely different.

I'm still thinking if there are a better solution than using a shared_ptr? This design seems to be the actual cause for this issue.

Reply:
What we actually need is either an exclusive mutable borrow for the nested scope, or passing ownership into the nested scope and returning it back by moving. Replacing the copy with a move constructor and using a std::unique_ptr might work.

Reply:
i think one of the issue can be solved by renaming. 
ScopedPainter sounds like a ScopedPointer holding a Painter. 
This is not the case we have a PainterState valid for a scope. 

So how about creating a PainterState class holding the shared pointer painter.
This can than controlled by a scoped pointer or just live on the stack in a braces scope of function.

Reply:
I don't know how to achieve this. Can you elaborate a little bit more?

Reply:
My idea with the move constructor doesn't work.

Wrapping twice into a `SharedPainter` (std::shared_ptr) and a `SharedPainterScope` (save/restore) makes sense to separate responsibilities. I'm not sure if it is over-engineering in this case.

The std::shared_ptr still feels inappropriate for managing nested scope, but I don't know of any other solution right now. The usability issues when passing by const-ref vs. by-val is an indicator that the design doesn't really fit.

Reply:
Maybe we can just pass QPainter around and that's it. At least the code is split into functions and it's a little bit easy to understand.

Reply:
Maybe supported by a simple, non-copyable `PainterSaver` (like QMutexLocker) that takes care of the save() and restore() calls in it's ctor/dtor. It might expose a public restore() function that releases and returns the wrapped QPainter pointer.

Reply:
What about we go functional style?

```c++
class ScopedPainter {
  public:
    ScopedPainter() = delete;

    explicit ScopedPainter(QPaintDevice* pPaintDevice)
            : m_painter(pPaintDevice) {
    }

    void withNewScope(const std::function<void()>& operation) {
        m_painter.save();
        operation();
        m_painter.restore();
    }

    QPainter* operator->() {
        return &m_painter;
    }

  private:
    QPainter m_painter;
};
```

Then we can do:
```c++
scopedPainter.withNewScope([&](){
        scopedPainter->setOpacity(0.5);
        // We can even nest scopes
        scopedPainter.withNewScope([&](){
            scopedPainter->setOpacity(0.4);
            // And pass the painter with certain scope to a function
            drawSomething(scopedPainter);
        });
    });
```

Copy constructor is deleted so you can only capture by reference.

Reply:
I like the idea with the PainterSaver.
It is a known concept and probably produces easy to read code.

I wonder how we deal with the ownership. The painting function itself can take a raw pointer, promissing not to store it. If it uses the PainterSaver, which stores the pointer, the promise is still Ok, because PainterSaver lives on the stack. Unfortunately it stores the pointer ...

Reply:
The same applies to QMutexLocker, so why worry?